include/kernel.h

```diff
@@ -0,0 +1,534 @@
+extern kevent_t KEVENT_IRQ_DP;
+/** @brief Event generated when the RESET button is pushed (pre-NMI exception) */
+extern kevent_t KEVENT_RESET;
+
+
+/** 
+ * @brief Initialize the multi-threading kernel
+ *
+ * The current execution context becomes the main thread of the
+ * program, with priority set to 0 (you can change priority
+ * to the main thread as well using #thread_set_pri).
+ *
+ * The main thread uses the original stack allocated for the
+ * whole process, so it is technically unbounded (or limited by
+ * heap size).
+ *
+ * @return a pointer to the main thread.
+ *
+ */
+kthread_t* kernel_init(void);
+
+/** 
+ * @brief Shutdown the multi-threading kernel
+ *
+ * This function is mostly useful for testing purposes.
+ * Since the kernel does not keep track of all created threads,
+ * this function should be called only when all created threads
+ * are exited or have been killed.
+ */
+void kernel_close(void);
+
+/**
+ * @brief Create a new thread
+ *
+ * Create a new thread, with a specified stack size and priority.
+ * The thread is immediately made ready after creation, so if it has
+ * a priority higher than or equal to the current thread, it will be
+ * scheduled immediately, before thread_new returns.
+ *
+ * The thread will begin execution from the specified entry point
+ * function. If the function ever returns, the thread is automatically
+ * killed.
+ *
+ * @param[in] name
+ *            Name of the thread (for debugging purposes)
+ * @param[in] stack_size
+ *            Size of the stack in bytes. Minimum suggested size
+ *            is 2048.
+ * @param[in] pri
+ *            Priority of the thread (0-127). Negative values are not allowed.
+ *            Higher number means higher priority.
+ * @param[in] user_entry
+ *            Entry point of the thread. This is the function that will
+ *            be called when the thread begins execution. If this function
+ *            ever returns, the thread is automatically killed.
+ * @param[in] user_data
+ *            Argument that will be passed to the entry point of the thread.
+ *
+ * @return    A pointer to the new thread. It is not necessary to store
+ *            this reference if not required; the thread will clean up
+ *            after itself when it exits.
+ */
+kthread_t* kthread_new(const char *name, int stack_size, int8_t pri,
+	void (*user_entry)(void*), void *user_data);
+
+
+/**
+ * @brief Yield execution of the current thread and run the scheduler.
+ *
+ * This function allows the current thread to cooperatively yield
+ * its execution to allow other threads to run.
+ *
+ * The scheduler will switch to the highest priority thread that is
+ * currently ready to run. If no ready thread has a priority higher than
+ * the current thread, the scheduler will switch to a different thread 
+ * of the same priority of the current one (the scheduler will guarantee
+ * a correct round-robin scheduling among threads of the same priority).
```

Comment:
What about the case where the thread creates a new thread of the same priority? Will it get scheduled after all of them in case all the remaining threads are of same priority?

Reply:
This is answered in `kthread_new` documentation:

> * The thread is immediately made ready after creation, so if it has
> * a priority higher than or equal to the current thread, it will be
> * scheduled immediately, before thread_new returns.